### DISCUSSION TITLE
Cockpit APIs for programmatically managing systemd services

### DISCUSSION THREAD

**Discussion by prashanth057** [cockpit-project/cockpit#18281] [General]
Hello,
I am fairly new to cockpit and have been in exploratory mode checking out its features and capabilities. I have setup a local cockpit instance and have been playing around with the UI.
One of my main use cases  I have is to manage my application services which run as systemd service on a large cluster of nodes. I am primarily interested in using cockpit as an API which I could invoke to start/stop/check status of my application across 100s of hosts programmatically(I don't really need the GUI layer). 
Additionally while doing this I would also need to plugin custom authentication and authorisation layer to ensure only allowed users are able to do this function. Is this doable with cockpit?
If yes, can someone point me to the API documentation for the same?

**Comment by marusak** [comment]
What you are looking for is [Ansible](https://www.ansible.com/). It is exactly meant for your needs when you need to manage multiple hosts. It has [systemd module](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/systemd_module.html) for very easy work with systemd units.

Cockpit is really meant for deep dive into one machine to debug it in graphical interface.

**Comment by prashanth057** [comment]
Thanks for the reply. I am looking at Ansible as well.

Just to clarify my complete intent and why I thought cockpit may work - I actually intend to build an aggregated custom GUI which when logged in would show the status of the application across all the hosts using a backend service which could possibly use cockpit APIs to get the status of the systemd service.
The aggregated GUI will also need to show systemd service logs for any failed services and allow user to start/stop the systemd service.
The backend service powering the GUI could leverage cockpit APIs to do these.
Some features such as viewing realtime logs and realtime systemd service status looked more efficient with cockpit APIs since it uses a web socket rather than Ansible. I would think with Ansible I would need to keep calling status and log commands   Continuously to keep my aggregated GUi view upto date.

**Reply by marusak** [reply]
I will just repeat that Cockpit is not meant for fleet management and it does not scale well. There used to be `Dashboard` plugin that we eventually dropped and it was limited to 20 machines.

But to answer your question - yes, there is a way how to use Cockpit and how to manage systemd units, see https://github.com/cockpit-project/cockpit/blob/main/pkg/lib/service.js
